sound/soc/sof/control.c

```diff
@@ -244,6 +297,8 @@ int snd_sof_bytes_get(struct snd_kcontrol *kcontrol,
 	struct sof_abi_hdr *data = cdata->data;
 	size_t size;
 
+	snd_sof_refresh_control(scontrol);
```

Comment:
this makes sense for bytes controls, but volume/enum/switch controls too? Why will the firmware automatically mess with these values?

Reply:
We we're using enum with @singalsu for direction-of-arrival "angles" callback with beamformer. Could be done with bytes, but would be totally proprietary (unreadable) by normal people.

Anyway despite the above, to me it would seem appropriate just to refresh all control types. Can't imagine what harm it could do and we would be prepared for all possible future scenarios.

Reply:
@ranj063, I agree with @juimonen that having a generic implementation is better so in the future when SOF gains more autonomous features or adds support for different changes benefiting the user then we do not have to revisit the code.

I'm not sure how user space handles the current SOF_TPLG_KCTL_BYTES_VOLATILE_RO, but if they need to poll to notice a change then they can be switch to this notification method without the volatile mode.
Even if user space keeps polling the firmware and topology can be changed to take advantage of the fact that we no longer need to do IPC if the control is not changed on the SOF side..

Reply:
I think it's nice that this works for all controls. The existing basic components will remain kernel to FW direction. But we could have instead of fixed volume control an automatic gain control or level control with user (initial) gain setting. The gain changes done by algorithm would be visible to user space too. Switch is practical for binary audio events indication. I think this opens the space for a lot of innovation.

Reply:
@ranj063 There's platform power impact if the updates happen often so need to be careful and do internal filtering in FW to have e.g. thresholds and hysteresis to pass them to user space. The recommendation is that a component should have switch control (default off) to enable updates. I'm adding some examples to beamformer PR to show FW code and topology example.